contributors/design-proposals/auth/kubectl-exec-plugins.md

```diff
@@ -0,0 +1,264 @@
+to wrap `kubectl` so the tool is invoked on every command. For example, the following
+is a [real example](
+https://github.com/heptio/authenticator#4-set-up-kubectl-to-use-heptio-authenticator-for-aws-tokens)
+from Heptio's AWS authenticator:
+
+```
+kubectl --kubeconfig /path/to/kubeconfig --token "$(heptio-authenticator-aws token -i CLUSTER_ID -r ROLE_ARN)" [...]
+```
+
+Beside resulting in a long command, this potentially encourages distributions to
+wrap or fork kubectl, changing the way that users interact with different
+Kubernetes clusters.
+
+# Proposal
+
+This proposal builds off of earlier requests to [support exec-based plugins](
+https://github.com/kubernetes/kubernetes/issues/35530#issuecomment-256170024), and
+proposes that we should add this as a first-class feature of kubectl. Specifically,
+client-go should be able to receive credentials by executing a command and reading
+that command's stdout.
+
+In fact, client-go already does this today. The GCP plugin can already be configured
+to [call a command](
+https://github.com/kubernetes/client-go/blob/kubernetes-1.8.5/plugin/pkg/client/auth/gcp/gcp.go#L228-L240)
+other than `gcloud`.
+
+## Plugin responsibilities
+
+Plugins are exec'd through client-go and print credentials to stdout. Errors are
+surfaced through stderr and a non-zero exit code. client-go will use structured APIs
+to pass information to the plugin, and receive credentials from it.
+
+```go
+// ExecCredentials are credentials returned by the plugin.
+type ExecCredentials struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+    // Token is a bearer token used by the client for request authentication.
+    Token string `json:"token,omitempty"`
+    // Expiry indicates a unix time when the provided credentials expire.
+    Expiry int64 `json:"expiry,omitempty"`
+}
+
+// ExecInfo is structed information passed to the plugin.
+type ExecInfo struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+    // InvalidCredentials is true after the API returned a 401 HTTP status
+    // suggesting previous credentials were invalid.
+    InvalidCredentials bool `json:"invalidCredentials,omitempty"`
+}
+```
+
+To instruct client-go to use the bearer token `BEARER_TOKEN`, a plugin would print:
+
+```terminal
+$ ./kubectl-example-auth-plugin
+{
+  "kind": "ExecCredentials",
+  "apiVersion":"client.authentication.k8s.io/v1alpha1",
+  "token":"BEARER_TOKEN"
+}
+```
+
+To surface runtime-based information to the plugin, such as a request body for request
+signing, client-go will set the environment variable `KUBERNETES_EXEC_INFO` to a JSON
+serialized Kubernetes object when calling the plugin.
+
+
+```
+KUBERNETES_EXEC_INFO='{
+    "kind":"ExecInfo",
+    "apiVersion":"client.authentication.k8s.io/v1alpha1",
+    "invalidCredentials": true
```

Comment:
since the plan is for this to live in client-go, and be usable by other components (such as the kubelet, for obtaining initial client certificates), we need to make sure the provider knows whether the session is interactive (so it knows whether to prompt or just fail if input is needed). can that be detected by the plugin, or do we need to make that part of the input?